### ISSUE TITLE
[unified-signatures] Doesn't consider ordering and type collapsing when considering if a rewrite makes sense

### ISSUE THREAD

**Issue by RubyTunaley** [typescript-eslint/typescript-eslint#2793]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
	"parser": "@typescript-eslint/parser",
	"plugins": ["@typescript-eslint"],
	"rules": {
		"@typescript-eslint/unified-signatures": "warn"
	}
}
```

```TS
interface TypeA
{
	readonly kind: 'a';
}

interface TypeB
{
	readonly kind: 'b';
}

type TypeAB = TypeA | TypeB;

function foo (x: TypeA): boolean | null;
function foo (x: TypeB): false;
function foo (x: TypeAB): boolean | null;
function foo (x: TypeAB): boolean | null
{
	// Implementation doesn't matter
	return null;
}
```

**Expected Result**

No warning is generated.

**Actual Result**

A warning is generated on the 3rd overload of `foo` saying it can be merged with the first overload.

```
15:15  warning  This overload and the one on line 13 can be combined into one signature taking `TypeA | TypeAB`  @typescript-eslint/unified-signatures
```

**Additional Info**

TypeScript collapses `TypeA | TypeAB` to `TypeAB`, meaning that if the linting advice was taken the 2nd overload specifying a different return type for `TypeB` would never be reached. This could result in runtime errors.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.8.1` |
| `@typescript-eslint/parser`        | `4.8.1` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `15.0.1` |

**Comment by bradzacher** [comment]
The warning is 100% correct.
You can merge the two signatures and produce the exactly same result.
Try commenting out the signature in [the playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCoE8AOECCBYAKAG9CBIKCOAEwHsQAbdZAa1CoC5kByOLgbkIBfQoVCRYiFBmwAhQiQLlKtBk1YgO3AEb8hIgmCxSjOGcgC8aE8gA+V2QIKEYAVxAIwwOshg0ayAAoAD05pXABKTi0-ekoQW2QQF3p6R1d3T29ff2DQoxlInzh6AGcINLcPL3jswJD7XAKomLiEpJSKjOqfPzq87FNC6JpYuHi7dvp5MgB6GeQASQBbTFiliHA4TPjaCBKQLjBkJa3xMgowFyh4ycdhJweqCAR6OApkBDoSo-qwnEdPiBvsgglALD0aMFwo5CE8Xm8UIDgeh+hAZACvkd0GDLNkAuhofo4a93kijgAvVGmDFAik4iEBcmEh5AA) and you'll see the rule is right.

What it is wrong is the fix - the fix should merge line 13 into line 15, not line 15 into line 13.
However I believe that this is nearly impossible to detect reliably.
This ultimately comes down to the rule understanding developer intent, which is difficult to do.

**Comment by RubyTunaley** [comment]
Ah yes you're right, line 13 is redundant in this example, my bad.

If finding out the merge direction is off the table it might still be worth rewording the warning message to explicitly say to consider merging in both directions before suppressing the warning. For example:

```
15:15  warning  This overload and the one on line 13 can be combined into one signature taking `TypeA | TypeAB`; consider both possible overload orderings  @typescript-eslint/unified-signatures
```

**Comment by bradzacher** [comment]
As with everything. This project is open source and community maintained. Happy to accept a PR.

If fixing the direction entirely is something you're interested in - happy to accept a PR for that.

**Comment by arminyahya** [comment]
Hi everyone can i work on this waning message?
